internal/provider/resource_tfe_project_oauth_client_.go

```diff
@@ -0,0 +1,173 @@
+		Include: []tfe.OAuthClientIncludeOpt{tfe.OauthClientProjects},
+	})
+	if err != nil {
+		if errors.Is(err, tfe.ErrResourceNotFound) {
+			log.Printf("[DEBUG] Oauth client %s no longer exists", oauthClientID)
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("error reading configuration of oauth client %s: %w", oauthClientID, err)
+	}
+
+	isProjectAttached := false
+	for _, project := range oauthClient.Projects {
+		if project.ID == projectID {
+			isProjectAttached = true
+			d.Set("project_id", projectID)
+			break
+		}
+	}
+
+	if !isProjectAttached {
+		log.Printf("[DEBUG] Project %s not attached to oauth client %s. Removing from state.", projectID, oauthClientID)
+		d.SetId("")
+		return nil
+	}
+
+	d.Set("oauth_client_id", oauthClientID)
+	return nil
+}
+
+func resourceTFEProjectOauthClientDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(ConfiguredClient)
+
+	oauthClientID := d.Get("oauth_client_id").(string)
+	projectID := d.Get("project_id").(string)
+
+	log.Printf("[DEBUG] Detaching project (%s) from oauth client (%s)", projectID, oauthClientID)
+	oauthClientRemoveProjectsOptions := tfe.OAuthClientRemoveProjectsOptions{}
+	oauthClientRemoveProjectsOptions.Projects = append(oauthClientRemoveProjectsOptions.Projects, &tfe.Project{ID: projectID})
+
+	err := config.Client.OAuthClients.RemoveProjects(ctx, oauthClientID, oauthClientRemoveProjectsOptions)
+	if err != nil {
+		return fmt.Errorf(
+			"error detaching project %s from oauth client %s: %w", projectID, oauthClientID, err)
+	}
+
+	return nil
+}
+
+func resourceTFEProjectOauthClientImporter(ctx context.Context, d *schema.ResourceData, meta interface{}) ([]*schema.ResourceData, error) {
+	// The format of the import ID is <ORGANIZATION/PROJECT ID/OAUTHCLIENT NAME>
+	splitID := strings.SplitN(d.Id(), "/", 3)
+	if len(splitID) != 3 {
+		return nil, fmt.Errorf(
+			"invalid project oauth client input format: %s (expected <ORGANIZATION>/<PROJECT ID>/<OAUTHCLIENT NAME>)",
+			splitID,
+		)
+	}
+
+	organization, projectID, oauthClientName := splitID[0], splitID[1], splitID[2]
+
+	config := meta.(ConfiguredClient)
+
+	// Ensure the named project exists before fetching all the oauth clients in the org
+	_, err := config.Client.Projects.Read(ctx, projectID)
+	if err != nil {
+		return nil, fmt.Errorf("error reading configuration of project %s in organization %s: %w", projectID, organization, err)
+	}
+
+	options := &tfe.OAuthClientListOptions{Include: []tfe.OAuthClientIncludeOpt{tfe.OauthClientProjects}}
+	for {
+		list, err := config.Client.OAuthClients.List(ctx, organization, options)
+		if err != nil {
+			return nil, fmt.Errorf("error retrieving organization's list of oauth clients: %w", err)
+		}
+		for _, oauthClient := range list.Items {
+			if *oauthClient.Name != oauthClientName {
+				continue
+			}
+
+			for _, project := range oauthClient.Projects {
```

Comment:
To cut down request load while paginating through all OAuth clients in an org, if API supports, is it possible paginate through only OAuth clients belonging to the current ProjectID? See similar implementation in `data_source_project.go`

```
			readOptions := &tfe.WorkspaceListOptions{
				ProjectID: proj.ID,
			}

```

Reply:
I'm not completely sure because for policy sets, I did very similar thing to what I have here. The logic looks similar to data_source_project so I am not fully understanding the difference. Is this blocking or non-blocking comment?